### PR TITLE
STEP-1054: Use Test Repetition for Xcode 13 and above

### PR DESCRIPTION
**Pull request by Bence1001**
### Checklist

- [x] I've read and accepted the [Contribution Guidelines](https://github.com/bitrise-steplib/.github/blob/main/CONTRIBUTING.md)
- Requires _MINOR_ [version update](https://semver.org/)

### Context
* Xcode 13 supports Test Repetition to retry failing tests.
* This feature is already supported by the Xcode Test step. However, that reruns ALL tests, not just the ones that failed.

### Changes
**Use Xcode 13's Test Repetition feature if "Should retry test on failure?" is enabled.**
* This only works on Xcode 13+ stacks.
* Below Xcode 13, we keep using our own custom logic that reruns all tests in case of failure.
* Added E2E tests for this option (both enabled and disabled).
* Brokedown `runTest` function into smaller functions; extracted long parameter lists into structs.

### CODE REVIEW

File: xcodebuild.go [hunk 1]
```diff
@@ -209,70 +224,104 @@ func runTest(buildTestParams models.XcodeBuildTestParamsModel, outputTool, xcpre
...
+	return xcodebuildArgs, nil
+}
+
+func handleTestRunError(prevRunParams testRunParams, prevRunResult testRunResult) (string, int, error) {
+	if prevRunParams.swiftPackagesPath != "" && isStringFoundInOutput(cache.SwiftPackagesStateInvalid, prevRunResult.xcodebuildLog) {
+		log.RWarnf("xcode-test", "swift-packages-cache-invalid", nil, "swift packages cache is in an invalid state")
+		if err := os.RemoveAll(prevRunParams.swiftPackagesPath); err != nil {
+			log.Errorf("failed to remove Swift package caches, error: %s", err)
+			return prevRunResult.xcodebuildLog, prevRunResult.exitCode, prevRunResult.err
 		}
```
Comment (lpusok) [inline xcodebuild.go:261]: Should re-run in this case.

**Review by lpusok — APPROVED** [review]

**Review by lpusok — APPROVED** [review]

**Comment by Bence1001** [comment]
Checks are failing because the new test target is configured incorrectly in the sample. This PR should fix this problem: https://github.com/bitrise-io/sample-swift-project-with-parallel-ui-test/pull/5.

File: xcodebuild.go (hunk 1 above)
Comment (Bence1001) [reply xcodebuild.go:261]: This is the old version, which didn't rerun the tests either: https://github.com/bitrise-steplib/steps-xcode-test/pull/171/files/f197272d32e1fa0e03693942f4bd08de64d19c0e#diff-dccdd133d1820fcdb03d94b18332976edf6f3d73e9d73979a6d74c0bac37438bL146-L150.

I'm unfamiliar with the requirements/use cases here. Why should we rerun them?

Comment (lpusok) [reply xcodebuild.go:261]: The original intent was to re-run them in case the Swift Package Cache gets in an invalid state (we restore the cache and it may be too Old or something like that). I changed it as there is no reason to keep the current buggy behaviour.

File: e2e/bitrise.yml [hunk 2]
```diff
@@ -247,7 +425,7 @@ workflows:
         - verbose: "yes"
         - should_retry_test_on_fail: $RETRY_ON_FAILURE
         - cache_level: $CACHE_LEVEL
-        - collect_simulator_diagnostics: on_failure
+        - collect_simulator_diagnostics: $COLLECT_SIM_DIAGNOSTICS
```
Comment (godrei) [inline e2e/bitrise.yml:428]: what is the reason behind this change?

Comment (lpusok) [reply e2e/bitrise.yml:428]: There are tests that we except to fail (flaky test retries for example), and in that case the diagnostics are collected even if is just  a test failure not a test runner failure (We do not have an up-to-date logic to know which one is which kind of failure)

**Review by godrei — APPROVED** [review]

**Review by lpusok — APPROVED** [review]

**Comment by Bence1001** [comment]
Latest changes LGTM, merging this PR.
